libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigationProvider.kt

```diff
@@ -16,33 +16,48 @@ object MapboxNavigationProvider {
      * @param navigationOptions
      */
     @JvmStatic
+    @Synchronized
```

Comment:
does this really need to be `synchronized`? this suggests that it can and should be called from different threads. IMO we should advertise this should only be called from a single thread

synchronized is also something that should only be called from background threads, because it blocks the calling thread - meaning it will block the main thread

Reply:
Also `@Synchronized` is not thread safe for coroutines https://blog.danlew.net/2020/01/28/coroutines-and-java-synchronization-dont-mix/

Reply:
> synchronized is also something that should only be called from background threads
that's not true.  In `Singleton` pattern `synchronized` is used to prevent initialization from different threads. Either `main` or `worker`. `MapboxNavigationProvider` is an `object`, so kotlin compiler generates all `synchronized` methods under the hood. 
But we use that object to handle inner `MapboxNavigation`. And all accessors should be `synchronized` to have a valid state if a user calls it from different threads.

also there are no issues with coroutines. We don't have coroutines here and fun should be `suspend` to get issues with coroutines.

Reply:
> because it blocks the calling thread - meaning it will block the main thread

it's not true in general. It blocks calling thread only if the lock (monitor) is acquired by another thread. In that case `main` thread will be blocked until another thread finishes `Synchronized ` block. But it's better than getting an object with invalid state.

Reply:
> If we call it from a single thread it will never be blocked. It blocks calling thread only if the lock (monitor) is acquired by another thread. 

i agree with that statement. but the suggestion is that there is no need for us to make it synchronized, unless we are suggesting clients to access it from multiple threads.

why create a synchronizedList if you don't intend to use it on multiple threads?

Reply:
whenever i see synchronized, this is a suggestion to not call it on the UI thread. because if a background thread accesses it, it will block the UI thread.

Reply:
we can add docs to access it only from the `main` thread, but what if a user makes a mistake? 
if a user calls methods from different threads he will get hardly reproducible race bugs. It will be a pain to clarify and fix them.
what cons do you see if we tell users to use a main thread but still have `synchronized ` to handle mistakes?


> if a background thread accesses it, it will block the UI thread.

- if the operation is fast it's not a problem, the main thread will be unblocked very fast
- if the operation is slow we have problems anyway. If it runs on a worker thread, the main will be blocked too long. If it runs on the main, other code on the main will wait until the operation is finished. we may have ANR and freezes in both cases.

I'm not a big fan of `synchronized`, sometimes it's too tricky to use it, just want to cover all potential issues.

@Guardiola31337 , @LukasPaczos wdyt?